### PR TITLE
Cache nonparametric matrices

### PR DESCRIPTION
**Pull request by albi3ro**
Part of the reason for separating out the static method `compute_matrix` from the instance method `matrix` was the possibility of adding caching.  This PR adds the `lru_cache` to the `compute_matrix` and `compute_sparse_matrix` methods to improve performance of construction the matrices for large `Prod` and `Sum` classes.

This PR already seems to be showing nice performance boosts for sums and products, though further benchmarking will clarify the extent of the speedup.

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -75,6 +75,8 @@ keyword argument when using `GellMann`, which determines which of the 8 Gell-Man
 * Improve `qml.math.expand_matrix` method for sparse matrices.
   [(#3060)](https://github.com/PennyLaneAI/pennylane/pull/3060)
 
+* Adds caching to the `compute_matrix` and `compute_sparse_matrix` of simple non-parametric operations.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:78]:
```suggestion
* Adds caching to the `compute_matrix` and `compute_sparse_matrix` of simple non-parametric operations.
  [(#3134)](https://github.com/PennyLaneAI/pennylane/pull/3134)
```

**Comment by AmintorDusko** [comment]
Hi @albi3ro, could you please provide the benchmarks you already did?

**Comment by albi3ro** [comment]
@AmintorDusko 
Using https://gist.github.com/albi3ro/88630d00981847f4825e9bd3f8f23d2d

Screen Shot 2022-10-07 at 10 42 30 AM

There doesn't seem to be a difference in performance on dense matrices between the branch and master.

**Comment by AmintorDusko** [comment]
Hi @albi3ro, I checked the performance increase for this PR and something caught my attention. I will illustrate that with some micro benchmarks.
Since we are caching the process of constructing objects, the reason for observing a slight increase in performance for a scipy object (sparse matrix) and almost none for the dense objects (numpy array), is that we are building the sparse representation inefficiently. As we input (data, (row, col)) for building a CSR sparse array, scipy will build a coo matrix and convert it to csr, so we expend some time initializing objects and performing conversions. 
```python
def compute_sparse_Hadamard_matrix():
    return sparse.csr_matrix([[INV_SQRT2, INV_SQRT2], [INV_SQRT2, -INV_SQRT2]])

@lru_cache()
def compute_sparse_Hadamard_matrix_cache():
    return sparse.csr_matrix([[INV_SQRT2, INV_SQRT2], [INV_SQRT2, -INV_SQRT2]])

%%timeit
compute_sparse_Hadamard_matrix()
96.3 µs ± 410 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%%timeit
compute_sparse_Hadamard_matrix_cache()
30 ns ± 0.0771 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
If on the other hand, we initialize the CSR matrix in its own representation with (data, indices, indptr), we can observe a nice increase in performance and up to some degree a performance as good as its caching version. 
```python
def compute_sparse_Hadamard_matrix_csr():
    return sparse.csr_matrix((  [ INV_SQRT2,  INV_SQRT2,  INV_SQRT2, -INV_SQRT2], 
                                [0, 1, 0, 1], 
                                [0, 2, 4]), shape=(2, 2))

@lru_cache()
def compute_sparse_Hadamard_matrix_csr_cache():
    return sparse.csr_matrix((  [ INV_SQRT2,  INV_SQRT2,  INV_SQRT2, -INV_SQRT2], 
                                [0, 1, 0, 1], 
                                [0, 2, 4]), shape=(2, 2))

%%timeit
compute_sparse_Hadamard_matrix_csr
10.6 ns ± 0.0491 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)

%%timeit
compute_sparse_Hadamard_matrix_csr_cache
10.8 ns ± 0.0878 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)
```
I considered the cached and non-cached to be equivalent in this last example because both timings can fluctuate up to 1 ns.
For this same reason, we don't observe a significant increase for the dense representation, numpy builds it in an efficient way.

**Review by AmintorDusko — APPROVED** [review]
Hi @albi3ro, I checked the performance increase for this PR and found that we have a very nice gain in performance for the sparse operations with that. Some micro-benchmarks follow.
As we have now:
```python
def compute_sparse_Hadamard_matrix():
    return sparse.csr_matrix([[INV_SQRT2, INV_SQRT2], [INV_SQRT2, -INV_SQRT2]])

@lru_cache()
def compute_sparse_Hadamard_matrix_cache():
    return sparse.csr_matrix([[INV_SQRT2, INV_SQRT2], [INV_SQRT2, -INV_SQRT2]])

%%timeit
compute_sparse_Hadamard_matrix()
96.3 µs ± 410 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%%timeit
compute_sparse_Hadamard_matrix_cache()
30 ns ± 0.0771 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
I observe that we are building the sparse matrix representation inefficiently. As we input (data, (row, col)) for building a CSR sparse array, scipy will build a coo matrix and convert it to CSR, so we expend some time initializing objects and performing conversions. 
If we initialize the CSR matrix in its own representation with (data, indices, indptr), we can observe an increase in performance for the non-cached implementation, or last say, for first-time call in the cached version.  
```python
def compute_sparse_Hadamard_matrix_csr():
    return sparse.csr_matrix((  [ INV_SQRT2,  INV_SQRT2,  INV_SQRT2, -INV_SQRT2], 
                                [0, 1, 0, 1], 
                                [0, 2, 4]), shape=(2, 2))

@lru_cache()
def compute_sparse_Hadamard_matrix_csr_cache():
    return sparse.csr_matrix((  [ INV_SQRT2,  INV_SQRT2,  INV_SQRT2, -INV_SQRT2], 
                                [0, 1, 0, 1], 
                                [0, 2, 4]), shape=(2, 2))

%%timeit
compute_sparse_Hadamard_matrix_csr()
33.1 µs ± 452 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%%timeit
compute_sparse_Hadamard_matrix_csr_cache()
27.6 ns ± 0.216 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
As the main goal of this PR was to implement cache, and this showed a very nice performance boost for sparse operations, I consider this as work well done and I'm happy to approve it. Thank you for that.

**Pull request merged**